Optimize HTTP parsing

klaussfreire opened the pull request:
- Small cython code generation optimizations
- IOLoop optimization: Avoid heap maintenance when all timeouts in a small timeout queue are expired, something that tends to happen often
- Reduce reliance on regex matching, regexes are slow
- Implement HTTPHeaders from scratch, generating efficient
  code under Cython and avoiding all the inefficient abstraction
  of collections.MutableMapping

klaussfreire committed:
Small optimization of IOLoop cythonization

Make IOLoop a fast global variable

klaussfreire committed:
Small optimizations of tornado.gen

Avoid global references in the critical paths

klaussfreire committed:
HTTP parsing optimization

Avoid redundant attribute access and other code generation
improvements

klaussfreire committed:
IOLoop optimization

Avoid heap maintenance when all timeouts in a small timeout
queue are expired, something that tends to happen often

klaussfreire committed:
Optimize HTTP parsing

- Reduce reliance on regex matching, regexes are slow
- Implement HTTPHeaders from scratch, generating efficient
  code under Cython and avoiding all the inefficient abstraction
  of collections.MutableMapping
- Some other minor optimizations
- Changelog entry

klaussfreire committed:
Bump version tuple

llogiudice commented on tornado/httputil.py:
```diff
@@ -230,13 +239,127 @@ def parse(cls, headers):
         return h
 
-    # MutableMapping abstract method implementations.
+    def pop(self, key, default=__marker):
```
Shouldn't this method also update self._last_key ?

klaussfreire replied:
Why?

`_last_key` is only used for parsing headers, which happens with `add` exclusively, so I don't see a need

llogiudice replied:
The sequence I had in mind was something like:

- add (key) => self._last_key is set to key
- pop (key) => self._last_key not updated
- parse_line (line) => KeyError is raised because they key is no longer present

I'm not sure if that usage is valid (although I think it should), but it seems to me like it may present problems.

klaussfreire replied:
The original implementation with MutableDict also had that issue it would seem, so I don't think it's a problem

gefarion commented on tornado/gen.py:
```diff
@@ -293,6 +293,10 @@ def _make_coroutine_wrapper(func, replace_callback):
 
+    stack_context_state = stack_context._state
+    Return_ = Return
+    GeneratorType_ = GeneratorType
```
Why are you using this tmp? I think that GeneratorType_ is only used once.

klaussfreire replied:
Once... *per call*

gefarion commented on tornado/http1connection.py:
```diff
@@ -159,6 +159,10 @@ def read_response(self, delegate):
     @gen.coroutine
     def _read_message(self, delegate):
         need_delegate_close = False
+        if self.is_client:
```
Why not `is_client = self.is_client`?

klaussfreire replied:
This forces the local to be of type `bint` in Cython without having to declare it

gefarion commented on tornado/httputil.py:
```diff
@@ -230,13 +239,127 @@ def parse(cls, headers):
+        self._as_list.clear()
+        self._last_key = None
+
+    def update(*args, **kwds):
```
Why not `update(self, other = ()`?

klaussfreire replied:
That's a very good question. I copied this off of MutableDict.

gefarion commented on tornado/httputil.py:
```diff
@@ -230,13 +239,127 @@ def parse(cls, headers):
+        elif not args:
+            raise TypeError("update() takes at least 1 argument (0 given)")
+        self = args[0]
+        other = args[1] if len(args) >= 2 else ()
```
`len(args)` can't be > 2

klaussfreire replied:
I think not

gefarion commented on tornado/httputil.py:
```diff
@@ -230,13 +239,127 @@ def parse(cls, headers):
+    def get(self, key, default=None):
+        return self._dict.get(_normalized_headers[key], default)
+
+    def __contains__(self, key):
```
Don't you have to use `_normalized_headers`?

gefarion commented on tornado/httputil.py:
```diff
@@ -230,13 +239,127 @@ def parse(cls, headers):
+            return NotImplemented
+        return dict(self.items()) == dict(other.items())
+
+    def __ne__(self, other):
```
Is this not the default behavior for ne?

klaussfreire replied:
Seems not

llogiudice commented on tornado/ioloop.py:
```diff
@@ -1048,7 +1050,29 @@ def start(self):
+                            needs_heapify = True
+
+                    ntimeouts = len(timeouts)
+                    if ntimeouts > 0 and ntimeouts < 32:
```
What happens if 'ntimeouts' is greater than 32 and 'need_heapify' remains False because the previous branch wasn't taken ? Wouldn't we end up with an unheaped array of timeouts ? If so, I'd suggest setting 'needs_heapify' to True unconditionally, and only clearing it with this new fast check.

klaussfreire replied:
No, the timeouts array is already a heap. `needs_heapify` is only set (and only necessary) when the heap nature of it is destroyed somehow. In the above code, due to linear filtering of cancelled timeouts.

klaussfreire committed:
Some fixes in HTTPHeaders

klaussfreire committed:
Fix and improve HTTPHeaders.copy

klaussfreire committed:
Improve IOLoop timeout fast path

Remove the size limit for the timeout queue. Instead, check
both ends of the heap. Heaps are *roughly* sorted by deadline,
so if both ends are due, it's likely the optimization would
kick in, and we don't run a huge risk of iterating through
the list unnecessarily.

That also makes the optimization more useful, since it can
handle huge lists of due timeouts. Take the opportunity to
maintain timeout order by using an in-place sort. The algorithm
for in-place sort should still be better than extracting them
one at a time, which would amount to a rather inefficient
implementation of heap sort.

klaussfreire commented:
@tzulberti-jampp , any opinions?

tzulberti-jampp commented on tornado/http1connection.py:
```diff
@@ -244,34 +248,36 @@ def _read_message(self, delegate):
+                stream = self.stream
+                if stream is not None and not stream.closed():
+                    stream.set_close_callback(self._on_connection_close)
+                    stream = None
```
Why do you set the stream to `None` if this is going to be done on line 289?

klaussfreire replied:
To release the reference. You mean 281. That's after the yield, which can be an eternity later.

tzulberti-jampp commented on tornado/httputil.py:
```diff
@@ -230,13 +235,148 @@ def parse(cls, headers):
+          If key is not found, d is returned if given, otherwise KeyError is raised.
+        '''
+        try:
+            value = self._dict[_normalized_headers[key]]
```
Shouldn't this pop the value from `_as_list`?

klaussfreire replied:
See line 255
